src/core/net/srp_server.cpp

```diff
@@ -635,6 +637,7 @@ otError Server::ProcessHostDescriptionInstruction(Host &                   aHost
         Dns::ResourceRecord record;
 
         SuccessOrExit(error = Dns::Name::ReadName(aMessage, aOffset, name, sizeof(name)));
+        VerifyOrExit(Dns::Name::IsSubDomainOf(name, GetDomain()), error = OT_ERROR_PARSE);
```

Comment:
Is `OT_ERROR_PARSE` what we should indicate here? Typically `OT_ERROR_PARSE` is for malformed messages.

Reply:
Agree that `OT_ERROR_PARSE` is not a proper error code. The DNS response code for domain not matching is `kResponseNotZone` and `kResponseNotAuth`:
https://github.com/openthread/openthread/blob/c03458433ec188be21027d6fa0eab01d3dbf6c03/src/core/net/dns_headers.hpp#L267-L268

But we currently don't have an OT error code that is mapping to `kResponseNotZone`. For now, I think `OT_ERROR_SECURITY` may be a better error code as it will be converted to `kResponseRefused`:
https://github.com/openthread/openthread/blob/c03458433ec188be21027d6fa0eab01d3dbf6c03/src/core/net/srp_server.cpp#L71-L74
https://github.com/openthread/openthread/blob/c03458433ec188be21027d6fa0eab01d3dbf6c03/src/core/net/dns_headers.hpp#L263

One thing to notice is that the SRP client converts `kResponseNotZone` to `OT_ERROR_PARSE`, which I think is not perfect either.
https://github.com/openthread/openthread/blob/c03458433ec188be21027d6fa0eab01d3dbf6c03/include/openthread/srp_client.h#L173

For the server, one solution is that we can have those DNS processing method return `Dns::Response` but not OT error code. For the client, we can have it return both OT error code (for request and response processing errors, e.g. `OT_ERROR_TIMEOUT` and DNS response code (clearly indicates the error happens on the server side). thoughts? @jwhui @abtink

Reply:
That is the reason I think converting errors for RCODRs is not a scalable idea. 
In DNSSD server, I choose to set the RCODE directly without returning any error. The server only return error in a server failure case, e.g. no more buffers.